### PR TITLE
 ✨  allowing to log back into a study when the tab was closed 

### LINKED ISSUE #1952: Give feedback on the status of a study (open, opening, closing, closed)
It would also be nice to know what the status of that resource is. In case of a study if it's open or being closed.

_Originally posted by @odeimaiz in https://github.com/ITISFoundation/osparc-simcore/pull/1939#discussion_r520622999_

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
The backend maintain a Redis-based registry where all user-allocated resources are kept (for now only socket IDs and project IDs).
These resources are linked to the user so-called "client session id" which is given by the frontend as the Broswer Tab ID when the websocket connection is created (Socket ID).

FRom the backend side it is not possible to distinguish between a network disconnection and a tab-closed. Therefore we do use a TTL (time-to-live) field in the Redis registry of 900 seconds. During that time, the same browser tab would be able to re-open the same study in case that it reconnects.

In the case of a closed tab, re-opening a new tab creates a new Tab ID. Therefore the backend treats that as "the user opened another tab and is trying to open a study that is already opened in another tab". Therefore it appears locked, and the user must wait 15 minutes until it gets closed, and after that the user may open it again.

# Changes
- [x] Listing the projects now will check if the user project is already opened by that very same user. If yes, it will check if a socket ID exists. If yes, then the behavior will remain the same. If not, then the project will not be defined as Locked and the user can re-open it.
- [x] If the project is just in the process of closing (e.g. dynamic services are saving their state and closing). then it should not be allowed to re-enter the project until the process is completed.
- [x] if a user is exporting a project, that project is locked until the export process is completed
- [x] if a user is duplicating a project, that project is locked until the cloning is completed
- [x] notification to any user that shares that project are in. @odeimaiz will probably vastlymprove the frontend part.

Bonus:
- ```make show_endpoints``` shows the endpoint to some of the services on demand
- codecov badge updated
- improve some test reliability by insuring redis database is properly cleaned
- importer now has a  first draft of a  migration system setup (see [models.py](https://github.com/ITISFoundation/osparc-simcore/pull/2383/files#diff-a5b899ca8e7dc2402e9bbe363e8e144f3f0e8bea16f58560f56808dd8fec8b2f))
- 

## Related issue/s
fixes [#431](https://github.com/ITISFoundation/osparc-issues/issues/431)
fixes #1952

### CODE REVIEW

File: packages/models-library/src/models_library/projects_state.py [hunk 1]
```diff
@@ -29,14 +29,47 @@ class DataState(str, Enum):
...
 
 
+@unique
+class ProjectStatus(str, Enum):
+    CLOSED = "CLOSED"
+    CLOSING = "CLOSING"
+    CLONING = "CLONING"
+    EXPORTING = "EXPORTING"
+    OPENING = "OPENING"
+    OPENED = "OPENED"
```
Comment (GitHK) [inline packages/models-library/src/models_library/projects_state.py:39]: would you mind providing more context to these states?

File: api/specs/common/schemas/node-meta-v0.0.1-converted.yaml [hunk 2]
```diff
@@ -16,7 +16,7 @@ properties:
   key:
     type: string
     description: distinctive name for the node based on the docker registry path
-    pattern: ^(simcore)/(services)/(comp|dynamic)(/[\w/-]+)+$
+    pattern: ^(simcore)/(services)/(comp|dynamic)(/[^\s/]+)+$
```
Comment (GitHK) [inline api/specs/common/schemas/node-meta-v0.0.1-converted.yaml:19]: Why this change in the regex?

File: services/web/client/source/class/osparc/dashboard/StudyBrowserButtonItem.js [hunk 3]
```diff
@@ -510,10 +511,28 @@ qx.Class.define("osparc.dashboard.StudyBrowserButtonItem", {
...
+          break;
+        case "EXPORTING":
+          toolTipText = lockedBy + this.tr(" is exporting it...");
+          break;
+        case "OPENING":
+          toolTipText = lockedBy + this.tr(" is opening it...");
+          break;
+        case "OPENED":
+          toolTipText = lockedBy + this.tr(" is using it.");
+          break;
```
Comment (GitHK) [inline services/web/client/source/class/osparc/dashboard/StudyBrowserButtonItem.js:532]: if these messages are shown to the user, maybe we should have less simplistic and more friendly wording?

File: services/web/server/src/simcore_service_webserver/exporter/formatters/models.py [hunk 4]
```diff
@@ -97,6 +98,9 @@ def _validate_version(cls, v):
         return str(v)
 
 
+_MIGRATION_FLAGS = dict(pre=True, always=True)
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/exporter/formatters/models.py:101]: why put it here, we have these flags in the validators usually. Is this because it will be shared by multiple validators?

File: services/web/server/src/simcore_service_webserver/exporter/request_handlers.py [hunk 5]
```diff
@@ -38,24 +42,38 @@ async def export_project(request: web.Request):
...
+        async with await lock_project(
+            request.app,
+            project_uuid,
+            ProjectStatus.EXPORTING,
+            user_id,
+            await get_user_name(request.app, user_id),
+        ):
+            await retrieve_and_notify_project_locked_state(
+                user_id, project_uuid, request.app
+            )
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/exporter/request_handlers.py:54]: Why is this not done inside the context manager, maybe by providing an option to it?

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 6]
```diff
@@ -477,32 +528,213 @@ async def trigger_connected_service_retrieve(
...
-        is_locked: bool = len(usernames) > 0
-        if is_locked:
+        user_session_id_list: List[Tuple[int, str]] = await rt.find_users_of_resource(
+            PROJECT_ID_KEY, project_uuid
+        )
+    set_user_ids = {x for x, _ in user_session_id_list}
+
+    assert (
+        len(set_user_ids) <= 1
+    )  # nosec  # NOTE: A project can only be opened by one user in one tab at the moment
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:668]:
```suggestion
    )  # nosec  # NOTE: A project can only be opened by one user in one tab
```

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 7]
```diff
@@ -477,32 +528,213 @@ async def trigger_connected_service_retrieve(
...
+        value=True,
+        owner=Owner(user_id=list(set_user_ids)[0], **usernames[0]),
+        status=ProjectStatus.OPENED,
+    )
+
+
+async def get_project_states_for_user(
+    user_id: int, project_uuid: str, app: web.Application
+) -> ProjectState:
+    # for templates: the project is never locked and never opened. also the running state is always unknown
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:713]: is unknown always used for templates? If yes, maybe a better name?

File: services/web/server/tests/unit/with_dbs/10/test_resource_manager.py [hunk 8]
```diff
@@ -497,6 +512,9 @@ async def test_interactive_services_removed_per_project(
     mocked_director_api["stop_service"].reset_mock()
 
 
+@pytest.mark.xfail(
+    reason="it is currently not permitted to open the same project from 2 different tabs"
+)
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/10/test_resource_manager.py:517]: was this permitted before? If yes, are we disabling it? In that case I would change the test to see that it properly fails to open from 2 different tabs

**Review by GitHK — COMMENTED** [review]
Nice addition! A few comments and questions below.

**Review by odeimaiz — APPROVED** [review]
💪 Thanks for bringing the Study statuses!

**Review by pcrespov — CHANGES_REQUESTED** [review]

File: services/web/server/src/simcore_service_webserver/projects/project_lock.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+PROJECT_REDIS_LOCK_KEY: str = "project:{}"
+
+
+async def lock_project(
+    app: web.Application,
+    project_uuid: Union[str, ProjectID],
+    status: ProjectStatus,
+    user_id: int,
+    user_name: Dict[str, str],
+) -> aioredlock.Lock:
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/project_lock.py:21]: VERY nice! 
It is not obvious to me that this return a context managed object, i would add a little doc describing usage. Something like: create a context to keep project locked
```python
async with await lock_project(...):
    # do some stuff with project, nobody can change it
```

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 10]
```diff
@@ -184,19 +189,65 @@ async def remove_services_and_data():
...
+                await retrieve_and_notify_project_locked_state(
+                    user_id, project_uuid, app
+                )
+
+            # save the state if the user is not a guest. if we do not know we save in any case.
+            with suppress(director_exceptions.DirectorException):
+                # here director exceptions are suppressed. in case the service is not found to preserve old behavior
+                await director_api.stop_services(
+                    app=app,
+                    user_id=user_id,
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:227]: use pylance and check all the type mismatches.. ``user_id`` being ``int`` or ``str`` is a classic. We have it in so many places ... we should slowly get it right.

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 11]
```diff
@@ -477,32 +528,213 @@ async def trigger_connected_service_retrieve(
...
+        )
+
+
 async def _get_project_lock_state(
-    user_id: int, project_uuid: str, app: web.Application
+    user_id: int,
+    project_uuid: str,
+    app: web.Application,
 ) -> ProjectLocked:
+    """returns the lock state of a project
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:645]: why not create a test of any of these entries?

File: api/specs/common/schemas/node-meta-v0.0.1-converted.yaml (hunk 2 above)
Comment (sanderegg) [reply api/specs/common/schemas/node-meta-v0.0.1-converted.yaml:19]: well. to be honest I just ran the ```make openapi-specs``` recipe.. and did not double-check if what it produces is correct. I can have a look though.

File: services/web/client/source/class/osparc/dashboard/StudyBrowserButtonItem.js (hunk 3 above)
Comment (sanderegg) [reply services/web/client/source/class/osparc/dashboard/StudyBrowserButtonItem.js:532]: this will be handled by @odeimaiz in a separate PR.

File: services/web/server/src/simcore_service_webserver/exporter/request_handlers.py (hunk 5 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/exporter/request_handlers.py:54]: Yes, I'd like that too. I'll see if python import mechanism allows that cause this will create a circular import. I'll check it out.

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 6 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:668]: this is not meant to remain like this forever. we should allow that later.

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 7 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:713]: no. this is a running state. what is the running state of templates? they never run. if you prefer I could put not_started.

File: services/web/server/tests/unit/with_dbs/10/test_resource_manager.py (hunk 8 above)
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/10/test_resource_manager.py:517]: to answer your questions, yes it was doable from the backend side but prohibited from the frontend.
so now it's in phase.
that marking as xfail actually means that it checks that the test fails.

File: packages/pytest-simcore/src/pytest_simcore/redis_service.py [hunk 12]
```diff
@@ -49,7 +49,7 @@ async def redis_service(redis_config: RedisConfig, monkeypatch) -> RedisConfig:
     return redis_config
 
 
-@pytest.fixture(scope="module")
+@pytest.fixture(scope="function")
 async def redis_client(loop, redis_config: RedisConfig) -> Iterator[aioredis.Redis]:
     client = await aioredis.create_redis_pool(redis_config.dsn, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/pytest-simcore/src/pytest_simcore/redis_service.py:54]: yes it flushes all the keys.

File: services/web/server/src/simcore_service_webserver/exporter/formatters/models.py [hunk 13]
```diff
@@ -131,3 +135,24 @@ def new_instance_from_shuffled_data(
...
 
         return new_obj
+
+    # migration validators --------------------------------
+    # NOTE: these migration validator are necessary when the base Project class is modified
+    # this allows importing an older project to the newest state
+    @validator("state", **_MIGRATION_FLAGS)
+    @classmethod
+    def optional_project_state_added_locked_status(cls, v):
+        """{"state":Optional[{"locked": {"value": bool}}}]
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/exporter/formatters/models.py:145]: the import_export_import actually tests this.

File: api/specs/common/schemas/node-meta-v0.0.1-converted.yaml (hunk 2 above)
Comment (sanderegg) [reply api/specs/common/schemas/node-meta-v0.0.1-converted.yaml]: added a test to prevent this from happening in the future

File: services/web/server/src/simcore_service_webserver/exporter/formatters/models.py (hunk 4 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/exporter/formatters/models.py:101]: moved to where it is used. and yes this needs to be shared by multiple migration validators

File: Makefile [hunk 14]
```diff
@@ -189,11 +189,24 @@ endif
 
 define _show_endpoints
 # The following endpoints are available
-echo "http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):9081                                                                 - oSparc platform"
-echo "http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):18080/?pgsql=postgres&username=scu&db=simcoredb&ns=public  - Postgres DB"
-echo "http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):9000                                                                 - Portainer"
+set -o allexport; \
+source $(CURDIR)/.env; \
+set +o allexport; \
+seperator=------------------------------------------;\
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply Makefile:195]: done

File: Makefile [hunk 15]
```diff
@@ -189,11 +189,24 @@ endif
...
+seperator=$${seperator}$${seperator}$${seperator};\
+rows="%-80s| %22s| %12s| %12s\n";\
+TableWidth=140;\
+printf "%80s| %22s| %12s| %12s\n" Endpoint Name User Password;\
+printf "%.$${TableWidth}s\n" "$$seperator";\
+printf "$$rows" 'http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):9081' 'oSparc platform';\
+printf "$$rows" 'http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):18080/?pgsql=postgres&username=scu&db=simcoredb&ns=public' 'Postgres DB' $${POSTGRES_USER} $${POSTGRES_PASSWORD};\
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply Makefile:202]: done

File: Makefile [hunk 16]
```diff
@@ -189,11 +189,24 @@ endif
...
+printf "$$rows" 'http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):9000' Portainer admin adminadmin;\
+printf "$$rows" 'http://$(if $(IS_WSL2),$(get_my_ip),127.0.0.1):18081' Redis
 endef
 
+show_endpoints:
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply Makefile:207]: done

File: packages/models-library/src/models_library/projects_state.py [hunk 17]
```diff
@@ -29,14 +29,47 @@ class DataState(str, Enum):
...
+
 class ProjectLocked(BaseModel):
-    value: bool = Field(
-        ..., description="True if the project is locked by another user"
+    value: bool = Field(..., description="True if the project is locked")
+    owner: Optional[Owner] = Field(
+        None, description="If locked, the user that owns the lock"
     )
-    owner: Optional[Owner] = Field(None, description="The user that owns the lock")
+    status: ProjectStatus = Field(..., description="The status of the project")
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects_state.py:47]: added more validations. and testing.

File: packages/models-library/tests/test_projects_state.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+def test_projects_state_model_examples(model_cls, model_cls_examples):
+    for name, example in model_cls_examples.items():
+        print(name, ":", pformat(example))
+        model_instance = model_cls(**example)
+        assert model_instance, f"Failed with {name}"
+
+
+def test_project_locked_with_missing_owner_raises():
+    with pytest.raises(ValueError):
+        ProjectLocked(**{"value": True, "status": ProjectStatus.OPENED})
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/tests/test_projects_state.py:20]: done

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 10 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:227]: I agree with you but this will be replaced by @GitHK PR on the dynamic sidecar.

File: services/web/server/src/simcore_service_webserver/exporter/request_handlers.py (hunk 5 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/exporter/request_handlers.py:54]: will add it once py3.8 is in

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 11 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:673]: this is tested in integration tests at the moment.

File: packages/models-library/src/models_library/projects_state.py (hunk 1 above)
Comment (sanderegg) [reply packages/models-library/src/models_library/projects_state.py:39]: hmm not sure what you want:
CLOSED = the project is closed
CLOSING = the project is in the process of being closed
etc etc...

**Review by pcrespov — APPROVED** [review]

File: packages/models-library/src/models_library/projects_state.py (hunk 1 above)
Comment (GitHK) [reply packages/models-library/src/models_library/projects_state.py:39]: I wasn't that clear, I wanted more of a lifecycle kind of explanation. But maybe it is not necessary.

File: packages/models-library/tests/test_services.py [hunk 19]
```diff
@@ -178,3 +178,36 @@ def test_services_model_examples(model_cls, model_cls_examples):
...
+@pytest.mark.parametrize(
+    "python_regex_pattern, json_schema_file_name, json_schema_entry_paths",
+    [
+        (SERVICE_KEY_RE, "project-v0.0.1.json", ["key"]),
+        (VERSION_RE, "project-v0.0.1.json", ["version"]),
+        (VERSION_RE, "node-meta-v0.0.1.json", ["version"]),
+        (SERVICE_KEY_RE, "node-meta-v0.0.1.json", ["key"]),
+    ],
+)
+def test_regex_pattern_same_in_jsonschema_and_python(
```
Comment (GitHK) [inline packages/models-library/tests/test_services.py:192]: 👍

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 20]
```diff
@@ -184,19 +188,94 @@ async def remove_services_and_data():
...
 
-@observe(event="SIGNAL_PROJECT_CLOSE")
+async def retrieve_and_notify_project_locked_state(
+    user_id: int, project_uuid: str, app: web.Application
+):
+    project = await get_project_for_user(app, project_uuid, user_id, include_state=True)
+    await notify_project_state_update(app, project)
+
+
+# TODO: Once python 3.8 is in use this
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:198]:
```suggestion
import sys
if sys.version_info >= (3, 8):
    raise Exception("use below implementation in py3.8+")
```

this way you won't miss it

**Review by GitHK — APPROVED** [review]
look OK, please check below suggestion before merging

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 20 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:198]: that is a good point ... 
I enabled it in py3.8 branch
